### ISSUE TITLE
Add metrics to record time costs of write system call in wal

### ISSUE THREAD

**Issue by Akiqqqqqqq** [etcd-io/etcd#17615]
### What would you like to be added?

We would like to have a metric to record time costs of **_write system call_** in wal in a single encoder.encode call

### Why is this needed?

Since our etcd cluster can go unstable with **constantly leader changes**, we tried to find the cause. 

But we can only observe that boltdb commits were slow(goes up to 4s at a time), other than that, any other metrics related to heartbeat loss, like wal fsync or cpu usage(goroutine cpu starvation) was normal.(fsync is up to 70ms at most)

We tried to look into it deeper, we use _pg_test_fsync_ and _biolalency_ , _ext4slower_ _ext4dist_ tools to locate the problem, at last we found the problem was at **_write system call_**, while the **_fsync system call_** was ok, **_write system call_** was so slow. So there was a kernel bug in **_write system call_**. After we upgraded kernel, the problem was solved.
ref: https://lore.kernel.org/all/20220817172139.3141101-1-shakeelb@google.com/

But still we cannot identify the direct cause of leader change at the beginning, only because there is only one metric recording the delay of **_fsync_** by wal, but the delay of **_write_** by wal is not counted. 

To summarize, we need a metric to record time costs of write system call in wal in a single encoder.encode call

**Comment by tjungblu** [comment]
Sorry for being late to the party. Shall we also instrument the other parts of the encode block? I would also be interested in how long it would take to crc+marshal the data. Which should also help us to understand whether WAL writing is slow due to the CPU.

Also, please don't forget to update our docs at: https://etcd.io/docs/v3.5/metrics/#disk

**Comment by Akiqqqqqqq** [comment]
> Sorry for being late to the party. Shall we also instrument the other parts of the encode block? I would also be interested in how long it would take to crc+marshal the data. Which should also help us to understand whether WAL writing is slow due to the CPU.
> 
> I reckon a metric for the full cycle of appending a record would also be useful.
> 
> Also, please don't forget to update our docs at: https://etcd.io/docs/v3.5/metrics/#disk edit: we could also think about adding an alert in https://github.com/etcd-io/etcd/tree/main/contrib/mixin/alerts (cc @Elbehery)

yup, I think it's a topic worth discussing, that there are still lof of process within the process of handling `ready `from raft. If we zoom out from `PageWriter.Write()`, where I added metric this time, we can see `crc.Write(rec.Data)` in `encoder.encode()` (the one you mentioned). Then if we move on zooming out, we can see `pbutil.MustMarshal(e)` in `WAL.saveEntry()`, and at last we will see the whole process of dealing `ready `from raft. 
I'm not whether to add metrics to cover every process above, or just a few of them? So this time I only added metric to catch  `PageWriter.Write()` to keep the minimum changes to source code. 
welcome to discuss

**Comment by ahrtr** [comment]
> I reckon a metric for the full cycle of appending a record would also be useful.

Sounds good to me.

**Comment by Akiqqqqqqq** [comment]
> > I reckon a metric for the full cycle of appending a record would also be useful.
> 
> Sounds good to me.

how about adding metrics to record the whole `case rd := <-r.Ready()` time costs in _server/etcdserver/raft.go_ L177?

**Comment by fuweid** [comment]
Hi @Akiqqqqqqq etcd wal doesn't use O_SYNC to write data. based on your reference of patch, it's about cgroup part. would you mind share more detail about what you see for the `write` syscall? Thanks!

**Comment by Akiqqqqqqq** [comment]
> Hi @Akiqqqqqqq etcd wal doesn't use O_SYNC to write data. based on your reference of patch, it's about cgroup part. would you mind share more detail about what you see for the `write` syscall? Thanks!

here's report

### background
1. etcd leader changed
2. boltdb commit slow
![image](https://github.com/etcd-io/etcd/assets/51102894/685f6ab0-c494-4a9a-b8aa-7fb5cb4245c7)


### debug
1. use pg_test_fsync and biolatency for testing
![image](https://github.com/etcd-io/etcd/assets/51102894/71cf0cad-80e0-49a7-b721-e9fbf12e014a)
![image](https://github.com/etcd-io/etcd/assets/51102894/f11133e7-8b80-458d-8d36-b3b47037e08b)
result shows everything fine

2. Use ext4slower and ext4dist to identify latency spikes originates from the write system call.
![image](https://github.com/etcd-io/etcd/assets/51102894/bc0172b4-007f-45f5-8e1a-47e6fb254112)
![image](https://github.com/etcd-io/etcd/assets/51102894/2f1d4993-2199-46f2-bccc-1121a59c243e)


- ext4slower result confirmed that latency originates from within `ext4_file_write_iter`, using ftrace's function graph to detect paths that take longer than 512 ms to complete. 
- It can be observed that the latency is due to multiple calls to `io_schedule_timeout`, which generally occurs due to excessive dirty page scaling during dirty page flow control. Therefore, subsequent activation of the corresponding tracepoints is necessary to observe detailed dirty page statistics.
![image](https://github.com/etcd-io/etcd/assets/51102894/569d5e37-23f5-4c31-8a28-6f0d65711775)
![image](https://github.com/etcd-io/etcd/assets/51102894/4fe1131d-fdc5-47ba-9069-bc5fd472fc52)

### tracepoint

- Turn on the switches at /sys/kernel/debug/tracing/events/writeback/balance_dirty_pages/enable and /sys/kernel/debug/tracing/events/writeback/global_dirty_state/enable to observe dirty page statistics.
- The process for obtaining dirty page statistics is as follows:
  - Global dirty pages
  - cgroup dirty pages
  - Determine if sleep is necessary
- Conclusion
  - Global dirty page levels are normal
  - Prior to the io_schedule_timeout sleep, the level of dirty pages is extremely high, likely due to a negative number being converted to an unsigned type, causing the issue.
![image](https://github.com/etcd-io/etcd/assets/51102894/ca7358ed-6e4f-4df9-a162-0ca66564fb61)

**Comment by fuweid** [comment]
@Akiqqqqqqq Thanks for the detail. It's very useful! Still thinking about `rstat` to merge cgroup information. so, you're running etcd in pod or container, it's correct?

**Comment by ahrtr** [comment]
https://github.com/etcd-io/etcd/pull/17618#issuecomment-2028805199

**Comment by Akiqqqqqqq** [comment]
> @Akiqqqqqqq Thanks for the detail. It's very useful! Still thinking about `rstat` to merge cgroup information. so, you're running etcd in pod or container, it's correct?

Hi, sorry for the late reply. Actually we are not running in container nor pod, we just run the binary with systemd. Since I havent  looked into the kernel part carefully, is there any problem?

**Comment by fuweid** [comment]
Hi @Akiqqqqqqq , it makes senses because the bug is triggered if process is running in cgroup. Confirmed v5.15.0-63 version has this issue https://elixir.bootlin.com/linux/v5.15.64/source/include/linux/memcontrol.h#L967. I also reproduced it with 32 vcore + 64 GiB with etcd. Thanks for sharing this!

```bash
# tracer: nop
#
# entries-in-buffer/entries-written: 2/2   #P:32
#
#                                _-----=> irqs-off
#                               / _----=> need-resched
#                              | / _---=> hardirq/softirq
#                              || / _--=> preempt-depth
#                              ||| / _-=> migrate-disable
#                              |||| /     delay
#           TASK-PID     CPU#  |||||  TIMESTAMP  FUNCTION
#              | |         |   |||||     |         |
            etcd-10422   [006] .....  1364.988339: balance_dirty_pages: bdi 252:0: limit=3201557 setpoint=2795669 dirty=18446744073709551614 bdi_setpoint=2795668 bdi_dirty=912 dirty_ratelimit=102400 task_ratelimit=0 dirtied=1024 dirtied_pause=1024 paused=0 pause=156 period=156 think=65400 cgroup_ino=3058
            etcd-15806   [030] .....  1613.615432: balance_dirty_pages: bdi 252:0: limit=3201557 setpoint=2700106 dirty=18446744073709551615 bdi_setpoint=2700105 bdi_dirty=960 dirty_ratelimit=102400 task_ratelimit=0 dirtied=1024 dirtied_pause=1024 paused=0 pause=124 period=124 think=5228 cgroup_ino=3058
```

**Comment by ahrtr** [comment]
Closing since all done.
